File: solutions/python/isbn-verifier/1/isbn_verifier.py
def is_valid(isbn: str) -> bool:
    checksum = 0
    factor = 10
    for c in isbn:
        if factor == 0:
            return False
        if c == "-":
            continue
        if c.isdecimal():
            checksum += int(c) * factor
        elif factor == 1 and c == "X":
            checksum += 10
        else:
            return False
        factor -= 1
    return factor == 0 and checksum % 11 == 0

File: solutions/haskell/isbn-verifier/1/src/IsbnVerifier.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE LambdaCase #-}

module IsbnVerifier (isbn) where

import Data.Char (digitToInt, isDigit)
import Data.Function ((&))
import Data.Maybe (fromMaybe)
import Data.Traversable (for)

isbn :: String -> Bool
isbn input = length digits == 10 && checkSum
  where
    digits = filter (/= '-') input
    checkSum =
        ( do
            xs <-
                for (zip [10, 9 ..] digits) \case
                    (1, 'X') -> Just 10
                    (factor, c) | isDigit c -> Just $ factor * digitToInt c
                    _ -> Nothing
            pure $ sum xs `mod` 11 == 0
        )
            & fromMaybe False
